server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -193,49 +206,106 @@ private MergeResult partialReduce(MergeTask task,
-        private int numReducePhases;
+        private final TopDocsStats topDocsStats;
+        private volatile MergeResult mergeResult;
+        private volatile boolean hasPartialReduce;
+        private volatile int numReducePhases;
 
-        PendingMerges(int bufferSize, int trackTotalHitsUpTo) {
-            this.bufferSize = bufferSize;
+        PendingMerges(int batchReduceSize, int trackTotalHitsUpTo) {
+            this.batchReduceSize = batchReduceSize;
             this.topDocsStats = new TopDocsStats(trackTotalHitsUpTo);
-            this.buffer = new QuerySearchResult[bufferSize];
         }
 
-        public boolean hasFailure() {
+        @Override
+        public synchronized void close() {
+            assert hasPendingMerges() == false : "cannot close with partial reduce in-flight";
+            if (hasFailure()) {
+                assert circuitBreakerBytes == 0;
+                return;
+            }
+            assert circuitBreakerBytes >= 0;
+            circuitBreaker.addWithoutBreaking(-circuitBreakerBytes);
+            circuitBreakerBytes = 0;
+        }
+
+        synchronized Exception getFailure() {
+            return failure.get();
+        }
+
+        boolean hasFailure() {
             return failure.get() != null;
         }
 
-        public synchronized boolean hasPendingMerges() {
+        boolean hasPendingMerges() {
             return queue.isEmpty() == false || runningTask.get() != null;
         }
 
-        public synchronized void sortBuffer() {
-            if (index > 0) {
-                Arrays.sort(buffer, 0, index, Comparator.comparingInt(QuerySearchResult::getShardIndex));
+        void sortBuffer() {
+            if (buffer.size() > 0) {
+                Collections.sort(buffer, Comparator.comparingInt(QuerySearchResult::getShardIndex));
             }
         }
 
+        synchronized long addWithoutBreaking(long size) {
+            circuitBreaker.addWithoutBreaking(size);
+            circuitBreakerBytes += size;
+            maxAggsCurrentBufferSize = Math.max(maxAggsCurrentBufferSize, circuitBreakerBytes);
+            return circuitBreakerBytes;
+        }
+
+        synchronized long addEstimateAndMaybeBreak(long estimatedSize) {
+            circuitBreaker.addEstimateBytesAndMaybeBreak(estimatedSize, "<reduce_aggs>");
+            circuitBreakerBytes += estimatedSize;
+            maxAggsCurrentBufferSize = Math.max(maxAggsCurrentBufferSize, circuitBreakerBytes);
+            return circuitBreakerBytes;
+        }
+
+        /**
+         * Returns the size of the serialized aggregation that is contained in the
+         * provided {@link QuerySearchResult}.
+         */
+        long ramBytesUsedQueryResult(QuerySearchResult result) {
+            if (hasAggs == false) {
+                return 0;
+            }
+            return result.aggregations()
+                .asSerialized(InternalAggregations::readFrom, namedWriteableRegistry)
```

Comment:
The `aggregations` should always be serialized here, right? I wonder if it is worth asserting that or something.